src/scheduler/speculative_scheduler.cpp

```diff
@@ -0,0 +1,121 @@
+#include "speculative_scheduler.h"
+
+#include "common/logging.h"
+#include "engine/engine.h"
+#include "memory/block_manager.h"
+#include "request/request.h"
+#include "request/sequence.h"
+#include "scheduler/response_handler.h"
+#include "scheduler/scheduler_policy.h"
+
+namespace llm {
+
+constexpr uint64_t kStepSleepTimeMs = 10;
+
+SpeculativeScheduler::SpeculativeScheduler(const SchedulerConfig& config,
+                                           Engine* llm_engine,
+                                           Engine* ssm_engine)
+    : config_(config), llm_engine_(llm_engine), ssm_engine_(ssm_engine) {
+  GCHECK(llm_engine_ != nullptr);
+  llm_block_manager_ = llm_engine_->block_manager();
+  GCHECK(ssm_engine_ != nullptr);
+  ssm_block_manager_ = ssm_engine_->block_manager();
+
+  tokenizer_ = llm_engine_->tokenizer();
+  GCHECK(llm_block_manager_ != nullptr);
+  GCHECK(ssm_block_manager_ != nullptr);
+  GCHECK(tokenizer_ != nullptr);
+
+  response_handler_ = new ResponseHandler(llm_block_manager_,
+                                          tokenizer_.get());
+  scheduler_policy_ = new FCFSSchedulerPolicy(response_handler_,
+                                              llm_block_manager_);
```

Comment:
nit: sounds you need to construct a joint block manager to include both slm and llm, so that both small and large model have a same memory allocation.

Reply:
i suppose only the ssm's output token ids are shared, do we have to use shared block manager to control it?